graphite_api/render/datalib.py

```diff
@@ -78,84 +78,106 @@ def __repr__(self):
             self.name, self.start, self.end, self.step)
 
 
-# Data retrieval API
-def fetchData(requestContext, pathExpr):
+class DataStore(object):
+    """
+    Simple object to store results of multi fetches.
+    Also aids in looking up data by pathExpressions.
+    """
+    def __init__(self):
+        self.paths = defaultdict(set)
+        self.data = defaultdict(list)
+
+    def add_path(self, path_expr, path):
+        """
+        Adds the path to the pathExpression.
+        """
+        self.paths[path_expr].add(path)
+
+    def get_paths(self, path_expr):
+        """
+        Returns all paths found for path_expr
+        """
+        return sorted(self.paths[path_expr])
+
+    def add_data(self, path, time_info, data):
+        """
+        Stores data before it can be put into a time series
+        """
+        # Dont add if empty
+        if not nonempty(data):
+            for d in self.data[path]:
+                if nonempty(d['values']):
+                    return
+
+        # Add data to path
+        self.data[path].append({
+            'time_info': time_info,
+            'values': data
+        })
+
+    def get_series_list(self, path_expr):
+        series_list = []
+        for path in self.get_paths(path_expr):
+            for data in self.data.get(path):
+                start, end, step = data['time_info']
+                series = TimeSeries(path, start, end, step, data['values'])
+                series.pathExpression = path_expr
+                series_list.append(series)
+        return series_list
+
+
+def fetchData(requestContext, pathExprs):
     from ..app import app
-
-    seriesList = []
     startTime = int(epoch(requestContext['startTime']))
     endTime = int(epoch(requestContext['endTime']))
 
-    def _fetchData(pathExpr, startTime, endTime, requestContext, seriesList):
-        matching_nodes = app.store.find(pathExpr, startTime, endTime)
+    # Convert to list if given single path
+    if not isinstance(pathExprs, list):
+        pathExprs = [pathExprs]
+
+    data_store = DataStore()
+    multi_nodes = defaultdict(list)
+    single_nodes = []
 
-        # Group nodes that support multiple fetches
-        multi_nodes = defaultdict(list)
-        single_nodes = []
-        for node in matching_nodes:
+    # Group nodes that support multiple fetches
+    for pathExpr in pathExprs:
+        for node in app.store.find(pathExpr):
```

Comment:
I've had a another opportunity to track this down, and I think this may be one of the main reasons for the problem I'm seeing.

Before, this was:
```
matching_nodes = app.store.find(pathExpr, startTime, endTime)
````
Meaning that only matched nodes with data were rendered.  Now, *every* node is rendered.  I'll raise a PR.